File: v1/native/process.ts
/**
 * Low-level helpers for spawning and stopping shell commands.
 */

import { run } from "./run"

export type SpawnedProcess = ReturnType<typeof Bun.spawn>

function isProcessGone(detail: string): boolean {
  const lower = detail.toLowerCase()
  return (
    lower.includes("not found") ||
    lower.includes("not running") ||
    lower.includes("no such process") ||
    lower.includes("no matching")
  )
}

/**
 * Recursively kill a Unix process and its descendants.
 * Avoids process-group kills (negative PIDs), which can take down App Runner itself
 * when the child shares the parent's group.
 */
async function killUnixTree(pid: number): Promise<void> {
  const children = await run(["pgrep", "-P", String(pid)])
  if (children.code === 0) {
    for (const line of children.stdout.split("\n")) {
      const childPid = Number(line.trim())
      if (Number.isInteger(childPid) && childPid > 0) {
        try {
          await killUnixTree(childPid)
        } catch {
          // child may have exited while we walked the tree
        }
      }
    }
  }
  const result = await run(["kill", "-9", String(pid)])
  if (result.code !== 0) {
    const detail = result.stderr.trim() || result.stdout.trim()
    if (!isProcessGone(detail)) {
      throw new Error(detail || `kill failed for pid ${pid}`)
    }
  }
}

/**
 * Force-kill an OS process by PID (and its descendants).
 * Used by app stop and the ports dialog.
 */
export async function killPid(pid: number): Promise<void> {
  if (!Number.isInteger(pid) || pid <= 0) {
    throw new Error("Invalid pid")
  }
  if (pid === process.pid) {
    throw new Error("Refusing to kill the Workspace Manager process")
  }

  if (process.platform === "win32") {
    const result = await run(["taskkill", "/PID", String(pid), "/T", "/F"])
    if (result.code !== 0) {
      const detail = result.stderr.trim() || result.stdout.trim()
      if (isProcessGone(detail)) return
      throw new Error(detail || `taskkill failed for pid ${pid}`)
    }
    return
  }

  await killUnixTree(pid)
}

/**
 * Best-effort kill of a spawned shell and its process tree.
 * Plain `child.kill()` only terminates cmd/sh, leaving npm/node/etc. running.
 */
export async function killProcess(child: SpawnedProcess) {
  const pid = child.pid
  if (typeof pid === "number" && pid > 0) {
    try {
      await killPid(pid)
      return
    } catch {
      // fall through — process may already be exiting
    }
  }
  try {
    child.kill()
  } catch {
    // already dead
  }
}

/**
 * Build a child env from the host process env plus app overrides.
 * Sets PYTHONUNBUFFERED and a default FORCE_COLOR for CLI friendliness.
 */
export function mergeSpawnEnv(
  appEnv: Record<string, string>
): Record<string, string> {
  const env: Record<string, string> = {}
  for (const [key, value] of Object.entries(process.env)) {
    if (typeof value === "string") env[key] = value
  }
  Object.assign(env, appEnv)
  env.PYTHONUNBUFFERED = "1"
  env.FORCE_COLOR = env.FORCE_COLOR ?? "0"
  env.NO_COLOR = env.NO_COLOR ?? "1"
  return env
}

export type SpawnShellOptions = {
  command: string
  cwd: string
  env: Record<string, string>
}

/**
 * Spawn a user/shell command via `cmd /c` (Windows) or `sh -c` (Unix).
 * stdout/stderr are piped; stdin is ignored.
 */
export function spawnShell(options: SpawnShellOptions): SpawnedProcess {
  const isWin = process.platform === "win32"
  return Bun.spawn({
    cmd: isWin
      ? ["cmd", "/c", options.command]
      : ["sh", "-c", options.command],
    cwd: options.cwd,
    env: options.env,
    stdout: "pipe",
    stderr: "pipe",
    stdin: "ignore",
  })
}

File: wails_backend/internal/native/process.go
package native

import (
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"strings"
)

func isProcessGone(detail string) bool {
	lower := strings.ToLower(detail)
	return strings.Contains(lower, "not found") ||
		strings.Contains(lower, "not running") ||
		strings.Contains(lower, "no such process") ||
		strings.Contains(lower, "no matching")
}

func killUnixTree(pid int) error {
	children, _ := Run([]string{"pgrep", "-P", strconv.Itoa(pid)})
	if children.Code == 0 {
		for _, line := range strings.Split(children.Stdout, "\n") {
			childPid, err := strconv.Atoi(strings.TrimSpace(line))
			if err != nil || childPid <= 0 {
				continue
			}
			_ = killUnixTree(childPid)
		}
	}
	result, err := Run([]string{"kill", "-9", strconv.Itoa(pid)})
	if err != nil {
		return err
	}
	if result.Code != 0 {
		detail := strings.TrimSpace(result.Stderr)
		if detail == "" {
			detail = strings.TrimSpace(result.Stdout)
		}
		if !isProcessGone(detail) {
			if detail == "" {
				detail = fmt.Sprintf("kill failed for pid %d", pid)
			}
			return fmt.Errorf("%s", detail)
		}
	}
	return nil
}

func KillPid(pid int) error {
	if pid <= 0 {
		return fmt.Errorf("Invalid pid")
	}
	if pid == os.Getpid() {
		return fmt.Errorf("Refusing to kill the Workspace Manager process")
	}
	if isWindows() {
		result, err := Run([]string{"taskkill", "/PID", strconv.Itoa(pid), "/T", "/F"})
		if err != nil {
			return err
		}
		if result.Code != 0 {
			detail := strings.TrimSpace(result.Stderr)
			if detail == "" {
				detail = strings.TrimSpace(result.Stdout)
			}
			if isProcessGone(detail) {
				return nil
			}
			if detail == "" {
				detail = fmt.Sprintf("taskkill failed for pid %d", pid)
			}
			return fmt.Errorf("%s", detail)
		}
		return nil
	}
	return killUnixTree(pid)
}

func MergeSpawnEnv(appEnv map[string]string) []string {
	env := os.Environ()
	seen := map[string]int{}
	for i, kv := range env {
		if eq := strings.IndexByte(kv, '='); eq > 0 {
			seen[strings.ToUpper(kv[:eq])] = i
		}
	}
	set := func(key, value string) {
		entry := key + "=" + value
		k := strings.ToUpper(key)
		if i, ok := seen[k]; ok {
			env[i] = entry
			return
		}
		seen[k] = len(env)
		env = append(env, entry)
	}
	for k, v := range appEnv {
		set(k, v)
	}
	set("PYTHONUNBUFFERED", "1")
	if _, ok := appEnv["FORCE_COLOR"]; !ok {
		if os.Getenv("FORCE_COLOR") == "" {
			set("FORCE_COLOR", "0")
		}
	}
	if _, ok := appEnv["NO_COLOR"]; !ok {
		if os.Getenv("NO_COLOR") == "" {
			set("NO_COLOR", "1")
		}
	}
	return env
}

func SpawnShell(command, cwd string, env []string) (*exec.Cmd, error) {
	var c *exec.Cmd
	if isWindows() {
		c = exec.Command("cmd", "/c", command)
	} else {
		c = exec.Command("sh", "-c", command)
	}
	c.Dir = cwd
	c.Env = env
	hideWindow(c)
	return c, nil
}
